File: src/types/custom_apps.ts
export type InstallerType = "nsis" | "innoSetup" | "msi" | "exe" | "unknown";

export interface CustomAppEntry {
  id: string;
  name: string;
  description: string | null;
  downloadUrl: string;
  installerType: InstallerType;
  silentArgs: string;
  expectedHash: string | null;
  createdAt: string;
  lastUsed: string | null;
}

export interface DownloadProgress {
  downloaded: number;
  total: number;
  filename: string;
}

export const INSTALLER_TYPE_LABELS: Record<InstallerType, string> = {
  nsis: "NSIS",
  innoSetup: "Inno Setup",
  msi: "MSI",
  exe: "Generic EXE",
  unknown: "Unknown",
};

export const INSTALLER_DEFAULT_ARGS: Record<InstallerType, string> = {
  nsis: "/S",
  innoSetup: "/VERYSILENT /SUPPRESSMSGBOXES /NORESTART",
  msi: "/qn /norestart",
  exe: "",
  unknown: "",
};

export function detectInstallerType(url: string): InstallerType {
  const filename = url.split("/").pop()?.toLowerCase() ?? "";
  if (filename.endsWith(".msi")) return "msi";
  if (filename.includes("setup") || filename.includes("install")) return "nsis";
  return "unknown";
}

File: src-tauri/src/models/custom_apps.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CustomAppEntry {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub download_url: String,
    pub installer_type: InstallerType,
    pub silent_args: String,
    pub expected_hash: Option<String>,
    pub created_at: String,
    pub last_used: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum InstallerType {
    Nsis,
    InnoSetup,
    Msi,
    Exe,
    Unknown,
}

#[allow(dead_code)]
impl InstallerType {
    pub fn default_args(&self) -> &str {
        match self {
            Self::Nsis => "/S",
            Self::InnoSetup => "/VERYSILENT /SUPPRESSMSGBOXES /NORESTART",
            Self::Msi => "/qn /norestart",
            Self::Exe | Self::Unknown => "",
        }
    }

    pub fn from_filename(filename: &str) -> Self {
        let lower = filename.to_lowercase();
        if lower.ends_with(".msi") {
            return Self::Msi;
        }
        if lower.contains("setup") || lower.contains("install") {
            return Self::Nsis;
        }
        Self::Unknown
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DownloadProgress {
    pub downloaded: u64,
    pub total: u64,
    pub filename: String,
}
